apps/wallet/src/background/accounts/zk/ZkAccount.ts

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+	type SerializedSignature,
+	type ExportedKeypair,
+	SIGNATURE_SCHEME_TO_FLAG,
+	toSerializedSignature,
+} from '@mysten/sui.js/cryptography';
+import { fromB64, toB64 } from '@mysten/sui.js/utils';
+import { blake2b } from '@noble/hashes/blake2b';
+import { toBigIntBE } from 'bigint-buffer';
+import { decodeJwt } from 'jose';
+import { bcs } from './bcs';
+import { getCurrentEpoch } from './current-epoch';
+import { type ZkProvider } from './providers';
+import {
+	type PartialZkSignature,
+	createPartialZKSignature,
+	fetchPin,
+	getAddress,
+	prepareZKLogin,
+	zkLogin,
+} from './utils';
+import {
+	Account,
+	type SerializedUIAccount,
+	type SigningAccount,
+	type SerializedAccount,
+} from '../Account';
+import networkEnv from '_src/background/NetworkEnv';
+import { type NetworkEnvType } from '_src/shared/api-env';
+import { deobfuscate, obfuscate } from '_src/shared/cryptography/keystore';
+import { fromExportedKeypair } from '_src/shared/utils/from-exported-keypair';
+
+type SessionStorageData = {
+	ephemeralKeyPair: ExportedKeypair;
+	proofs: PartialZkSignature;
+	minEpoch: number;
+	maxEpoch: number;
+	network: NetworkEnvType;
+};
+
+type JwtSerializedClaims = {
+	email: string;
+	fullName: string | null;
+	firstName: string | null;
+	lastName: string | null;
+	picture: string | null;
+	aud: string;
+	iss: string;
+	sub: string;
+};
+
+export interface ZkAccountSerialized extends SerializedAccount {
+	type: 'zk';
+	provider: ZkProvider;
+	/**
+	 * the pin used to create the account obfuscated
+	 */
+	pin: string;
+	/**
+	 * obfuscated data that contains user info as it was in jwt
+	 */
+	claims: string;
+}
+
+export interface ZkAccountSerializedUI extends SerializedUIAccount {
+	type: 'zk';
+	email: string;
+	picture: string | null;
+	fullName: string | null;
+}
+
+export function isZkAccountSerializedUI(
+	account: SerializedUIAccount,
+): account is ZkAccountSerializedUI {
+	return account.type === 'zk';
+}
+
+export class ZkAccount
+	extends Account<ZkAccountSerialized, SessionStorageData>
+	implements SigningAccount
+{
+	readonly canSign = true;
+	readonly unlockType = 'password' as const;
+
+	static async createNew({
+		provider,
+	}: {
+		provider: ZkProvider;
+	}): Promise<Omit<ZkAccountSerialized, 'id'>> {
+		const jwt = await zkLogin({ provider, prompt: 'select_account' });
+		const { pin } = await fetchPin(jwt);
+		const decodedJWT = decodeJwt(jwt);
+		if (
+			!decodedJWT.sub ||
+			!decodedJWT.iss ||
+			!decodedJWT.aud ||
+			!decodedJWT.email ||
+			typeof decodedJWT.email !== 'string'
+		) {
+			throw new Error('Missing jwt data');
+		}
+		// based on jose typings aud can be an array
+		// we expect it to be string so here just doing a join if necessary
+		const aud = Array.isArray(decodedJWT.aud) ? decodedJWT.aud.join(' ') : decodedJWT.aud;
```

Comment:
@mskd12 is this fine? what's the best way to handle this?

Reply:
The circuits are not designed to handle an array of aud values. Perhaps throw an error?

Reply:
Updated it [here](https://github.com/MystenLabs/sui/pull/13384)!